{
  "metadata": {
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3",
      "language": "python"
    },
    "language_info": {
      "name": "python",
      "version": "3.11.13",
      "mimetype": "text/x-python",
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "pygments_lexer": "ipython3",
      "nbconvert_exporter": "python",
      "file_extension": ".py"
    },
    "colab": {
      "provenance": [],
      "gpuType": "T4"
    },
    "accelerator": "GPU",
    "kaggle": {
      "accelerator": "none",
      "dataSources": [
        {
          "sourceId": 116062,
          "databundleVersionId": 14084779,
          "sourceType": "competition"
        }
      ],
      "dockerImageVersionId": 31192,
      "isInternetEnabled": true,
      "language": "python",
      "sourceType": "notebook",
      "isGpuEnabled": false
    }
  },
  "nbformat_minor": 0,
  "nbformat": 4,
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython\n",
        "!pip install obonet"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:11.489632Z",
          "iopub.execute_input": "2025-11-11T01:34:11.489927Z",
          "iopub.status.idle": "2025-11-11T01:34:24.935036Z",
          "shell.execute_reply.started": "2025-11-11T01:34:11.489902Z",
          "shell.execute_reply": "2025-11-11T01:34:24.933488Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EvPMdJS1jJgI",
        "outputId": "47022b1f-833a-4ef3-b04d-d759344a7b2c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/3.2 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m110.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.86\n",
            "Collecting obonet\n",
            "  Downloading obonet-1.1.1-py3-none-any.whl.metadata (6.7 kB)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.12/dist-packages (from obonet) (3.6.1)\n",
            "Downloading obonet-1.1.1-py3-none-any.whl (9.2 kB)\n",
            "Installing collected packages: obonet\n",
            "Successfully installed obonet-1.1.1\n"
          ]
        }
      ],
      "execution_count": 1
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import networkx as nx\n",
        "import Bio # Biopython library\n",
        "from Bio import SeqIO # SeqIO from Biopython for parsing sequence files\n",
        "import obonet # obonet for parsing OBO files\n",
        "import gc # garbage collection\n",
        "from sklearn.decomposition import PCA\n",
        "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay # metrics for model evaluation\n",
        "from sklearn.preprocessing import LabelEncoder # LabelEncoder for encoding categorical labels\n",
        "from sklearn.model_selection import train_test_split # for splitting data into training and testing sets\n",
        "import random"
      ],
      "metadata": {
        "id": "GGPDDs_kXqU9",
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:54.640598Z",
          "iopub.execute_input": "2025-11-11T01:34:54.640931Z",
          "iopub.status.idle": "2025-11-11T01:34:54.647575Z",
          "shell.execute_reply.started": "2025-11-11T01:34:54.640910Z",
          "shell.execute_reply": "2025-11-11T01:34:54.646336Z"
        }
      },
      "outputs": [],
      "execution_count": 2
    },
    {
      "cell_type": "code",
      "source": [
        "obo_path = '/kaggle/input/cafa-6-protein-function-prediction/Train/go-basic.obo' # Gene Ontology Basic OBO file\n",
        "fasta_path = '/kaggle/input/cafa-6-protein-function-prediction/Train/train_sequences.fasta'\n",
        "term_path = '/kaggle/input/cafa-6-protein-function-prediction/Train/train_terms.tsv'\n",
        "taxonomy_path = '/kaggle/input/cafa-6-protein-function-prediction/Train/train_taxonomy.tsv'"
      ],
      "metadata": {
        "id": "R7rUQCNdVug8",
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:54.910201Z",
          "iopub.execute_input": "2025-11-11T01:34:54.910560Z",
          "iopub.status.idle": "2025-11-11T01:34:54.916476Z",
          "shell.execute_reply.started": "2025-11-11T01:34:54.910536Z",
          "shell.execute_reply": "2025-11-11T01:34:54.914853Z"
        }
      },
      "outputs": [],
      "execution_count": 3
    },
    {
      "cell_type": "code",
      "source": [
        "term_df = pd.read_csv(term_path, sep='\\t')"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:55.809158Z",
          "iopub.execute_input": "2025-11-11T01:34:55.809539Z",
          "iopub.status.idle": "2025-11-11T01:34:56.223716Z",
          "shell.execute_reply.started": "2025-11-11T01:34:55.809514Z",
          "shell.execute_reply": "2025-11-11T01:34:56.222452Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 311
        },
        "id": "9I35MamqjJgM",
        "outputId": "bd93155b-347b-4efa-f0d9-a77bae207ed3"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: '/kaggle/input/cafa-6-protein-function-prediction/Train/train_terms.tsv'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3301142503.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mterm_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mterm_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1024\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1025\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1026\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1027\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1028\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    618\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    619\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 620\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    621\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    622\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1618\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1619\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1620\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1621\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1622\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1879\u001b[0m                     \u001b[0mmode\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m\"b\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1880\u001b[0;31m             self.handles = get_handle(\n\u001b[0m\u001b[1;32m   1881\u001b[0m                 \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1882\u001b[0m                 \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    871\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    872\u001b[0m             \u001b[0;31m# Encoding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 873\u001b[0;31m             handle = open(\n\u001b[0m\u001b[1;32m    874\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    875\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/kaggle/input/cafa-6-protein-function-prediction/Train/train_terms.tsv'"
          ]
        }
      ],
      "execution_count": 4
    },
    {
      "cell_type": "code",
      "source": [
        "taxonomy_df = pd.read_csv(taxonomy_path, sep='\\t', names=['EntryID', 'taxonomyID'])"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:58.231462Z",
          "iopub.execute_input": "2025-11-11T01:34:58.231834Z",
          "iopub.status.idle": "2025-11-11T01:34:58.308944Z",
          "shell.execute_reply.started": "2025-11-11T01:34:58.231808Z",
          "shell.execute_reply": "2025-11-11T01:34:58.307671Z"
        },
        "id": "dmMlIUYqjJgM"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "taxonomy_df.head() # Display the first few rows of the taxonomy DataFrame"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:58.822241Z",
          "iopub.execute_input": "2025-11-11T01:34:58.823680Z",
          "iopub.status.idle": "2025-11-11T01:34:58.858990Z",
          "shell.execute_reply.started": "2025-11-11T01:34:58.823630Z",
          "shell.execute_reply": "2025-11-11T01:34:58.857879Z"
        },
        "id": "8lLzwuiyjJgM"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "fasta_data = list(SeqIO.parse(fasta_path, \"fasta\")) # Parse the FASTA file and store sequences as a list of SeqRecord objects"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:34:59.065065Z",
          "iopub.execute_input": "2025-11-11T01:34:59.065361Z",
          "iopub.status.idle": "2025-11-11T01:35:00.596045Z",
          "shell.execute_reply.started": "2025-11-11T01:34:59.065341Z",
          "shell.execute_reply": "2025-11-11T01:35:00.594584Z"
        },
        "id": "SohLJahRjJgM"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "def get_processed_fasta_df(fasta_data):\n",
        "    \"\"\"Processes FASTA data into a DataFrame with EntryID and fasta_sequence.\"\"\"\n",
        "    fasta_dict_list = []\n",
        "    for fasta_seq in fasta_data:\n",
        "        # Extract EntryID from the FASTA header and store sequence\n",
        "        entry = fasta_seq.id.split('|')[1]\n",
        "        fasta_dict_list.append({\n",
        "            \"EntryID\": entry,\n",
        "            \"fasta_sequence\": str(fasta_seq.seq)\n",
        "        })\n",
        "    return pd.DataFrame(fasta_dict_list)\n",
        "\n",
        "# Process the FASTA data into a DataFrame\n",
        "fasta_df = get_processed_fasta_df(fasta_data)\n",
        "\n",
        "# Merge term_df with fasta_df based on 'EntryID'\n",
        "merged_df = pd.merge(term_df, fasta_df, on=\"EntryID\", how='left')\n",
        "\n",
        "# Merge the result with taxonomy_df based on 'EntryID'\n",
        "merged_df = pd.merge(merged_df, taxonomy_df, on=\"EntryID\", how=\"left\")\n",
        "\n",
        "# Display the first few rows of the merged DataFrame\n",
        "merged_df.head()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:01.195827Z",
          "iopub.execute_input": "2025-11-11T01:35:01.196218Z",
          "iopub.status.idle": "2025-11-11T01:35:02.037560Z",
          "shell.execute_reply.started": "2025-11-11T01:35:01.196193Z",
          "shell.execute_reply": "2025-11-11T01:35:02.036515Z"
        },
        "id": "Pbml269QjJgM"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "# EDA"
      ],
      "metadata": {
        "id": "eisjJ0jdjJgN"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Counting subontologies"
      ],
      "metadata": {
        "id": "JDumBUDYjJgN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import seaborn as sns # for data visualization\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a countplot of the 'aspect' column in term_df, representing subontologies (C, F, P)\n",
        "sns.countplot(data=term_df, x='aspect', palette='mako')\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:03.409558Z",
          "iopub.execute_input": "2025-11-11T01:35:03.410748Z",
          "iopub.status.idle": "2025-11-11T01:35:04.689640Z",
          "shell.execute_reply.started": "2025-11-11T01:35:03.410716Z",
          "shell.execute_reply": "2025-11-11T01:35:04.688325Z"
        },
        "id": "RvdVyvKvjJgO"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Counting terms for Cellular Component"
      ],
      "metadata": {
        "id": "J_OWfdEhjJgO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# number of most common terms to display\n",
        "k = 8\n",
        "\n",
        "# Get the k most common 'term' entries for 'Cellular Component' (aspect 'C')\n",
        "k_most_common_c_term_counts = term_df[term_df['aspect'] == 'C']['term'].value_counts().head(k)\n",
        "\n",
        "# Extract the term names from the counts for filtering and ordering\n",
        "k_most_common_c_term = [term for term in k_most_common_c_term_counts.index]\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a countplot for the top k cellular component terms\n",
        "# Filter the DataFrame to include only these terms\n",
        "sns.countplot(data=term_df[term_df['term'].isin(k_most_common_c_term)], x='term', palette='mako', order=k_most_common_c_term_counts.index)\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:04.855015Z",
          "iopub.execute_input": "2025-11-11T01:35:04.855702Z",
          "iopub.status.idle": "2025-11-11T01:35:05.208204Z",
          "shell.execute_reply.started": "2025-11-11T01:35:04.855674Z",
          "shell.execute_reply": "2025-11-11T01:35:05.206680Z"
        },
        "id": "5ZCKTBCPjJgO"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Counting terms for Molecular Function"
      ],
      "metadata": {
        "id": "9Z5ShHASjJgP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the k most common 'term' entries for 'Molecular Function' (aspect 'F')\n",
        "k_most_common_f_term_counts = term_df[term_df['aspect'] == 'F']['term'].value_counts().head(k)\n",
        "\n",
        "# Extract the term names from the counts for filtering and ordering\n",
        "k_most_common_f_term = [term for term in k_most_common_f_term_counts.index]\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a countplot for the top k molecular function terms\n",
        "# Filter the DataFrame to include only these terms\n",
        "sns.countplot(data=term_df[term_df['term'].isin(k_most_common_f_term)], x='term', palette='mako', order=k_most_common_f_term_counts.index)\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:06.104277Z",
          "iopub.execute_input": "2025-11-11T01:35:06.104660Z",
          "iopub.status.idle": "2025-11-11T01:35:06.433869Z",
          "shell.execute_reply.started": "2025-11-11T01:35:06.104637Z",
          "shell.execute_reply": "2025-11-11T01:35:06.432690Z"
        },
        "id": "IrBsTMJojJgP"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Counting terms for Biological Process"
      ],
      "metadata": {
        "id": "3L0SD5UmjJgP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the k most common 'term' entries for 'Biological Process' (aspect 'P')\n",
        "k_most_common_p_term_counts = term_df[term_df['aspect'] == 'P']['term'].value_counts().nlargest(k)\n",
        "\n",
        "# Extract the term names from the counts for filtering and ordering\n",
        "k_most_common_p_term = k_most_common_p_term_counts.index.tolist()\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a countplot for the top k biological process terms\n",
        "# Filter the DataFrame to include only these terms\n",
        "sns.countplot(data=term_df[term_df['term'].isin(k_most_common_p_term)], x='term', palette='mako', order=k_most_common_p_term_counts.index)\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:08.349480Z",
          "iopub.execute_input": "2025-11-11T01:35:08.349822Z",
          "iopub.status.idle": "2025-11-11T01:35:08.695975Z",
          "shell.execute_reply.started": "2025-11-11T01:35:08.349799Z",
          "shell.execute_reply": "2025-11-11T01:35:08.694821Z"
        },
        "id": "EV__fVDNjJgP"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Fasta sequence length histogram with kde,"
      ],
      "metadata": {
        "id": "DmhwdxsIjJgQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# number of sequences in fasta_data\n",
        "# len(fasta_data)\n",
        "\n",
        "# Extract fasta sequences as strings into a list\n",
        "fasta_sequences_list = [str(x.seq) for x in fasta_data]\n",
        "\n",
        "lengths = [len(seq) for seq in fasta_sequences_list]\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a histogram of sequence lengths with a Kernel Density Estimate (KDE)\n",
        "# 'kde=True' overlays a density curve\n",
        "sns.histplot(lengths, bins=500, kde=True, color='teal')\n",
        "\n",
        "plt.title('Distribution of Sequence Lengths')\n",
        "plt.xlabel('Sequence Length')\n",
        "plt.ylabel('Count')\n",
        "plt.grid(True, alpha=0.3)\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:13.163047Z",
          "iopub.execute_input": "2025-11-11T01:35:13.163456Z",
          "iopub.status.idle": "2025-11-11T01:35:14.546629Z",
          "shell.execute_reply.started": "2025-11-11T01:35:13.163431Z",
          "shell.execute_reply": "2025-11-11T01:35:14.545240Z"
        },
        "id": "_IyQ8tF6jJgQ"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Taxonomies countplot"
      ],
      "metadata": {
        "id": "CINbEounjJgQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the number of most common taxonomy IDs to display\n",
        "k = 8\n",
        "\n",
        "# Get the k most common 'taxonomyID' entries from the merged_df\n",
        "k_most_common_taxonomy_tid_counts = merged_df['taxonomyID'].value_counts().nlargest(k)\n",
        "\n",
        "# Extract the taxonomy IDs from the counts for filtering and ordering\n",
        "k_most_common_taxonomy_id = k_most_common_taxonomy_tid_counts.index.tolist()\n",
        "\n",
        "# Create a subset of the DataFrame containing only the top k taxonomy IDs\n",
        "subset_df = merged_df[merged_df['taxonomyID'].isin(k_most_common_taxonomy_id)]\n",
        "\n",
        "plt.figure(figsize=(10, 6))\n",
        "\n",
        "# Create a countplot for the top k taxonomy IDs\n",
        "sns.countplot(\n",
        "    data=subset_df,\n",
        "    x='taxonomyID',\n",
        "    order=k_most_common_taxonomy_tid_counts.index,\n",
        "    palette='mako'\n",
        ")\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:14.558742Z",
          "iopub.execute_input": "2025-11-11T01:35:14.559149Z",
          "iopub.status.idle": "2025-11-11T01:35:14.869385Z",
          "shell.execute_reply.started": "2025-11-11T01:35:14.559123Z",
          "shell.execute_reply": "2025-11-11T01:35:14.868384Z"
        },
        "id": "-sv7-mixjJgQ"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Amino acid counts"
      ],
      "metadata": {
        "id": "MTPPZvqCjJgQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from collections import Counter\n",
        "\n",
        "# Concatenate all fasta sequences into a single string\n",
        "all_sequences = \"\".join(merged_df['fasta_sequence'])\n",
        "\n",
        "# Count the occurrences of each amino acid in the combined sequence string\n",
        "aa_counts = Counter(all_sequences)\n",
        "\n",
        "# Convert the amino acid counts to a pandas DataFrame and sort by count in descending order\n",
        "aa_df = pd.DataFrame.from_dict(aa_counts, orient='index', columns=['count']).sort_values('count', ascending=False)"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:15.988964Z",
          "iopub.execute_input": "2025-11-11T01:35:15.989463Z",
          "iopub.status.idle": "2025-11-11T01:35:36.459077Z",
          "shell.execute_reply.started": "2025-11-11T01:35:15.989384Z",
          "shell.execute_reply": "2025-11-11T01:35:36.457510Z"
        },
        "id": "NnRHkTt6jJgR"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# Create a bar plot of amino acid counts\n",
        "# 'data' is the DataFrame of amino acid counts, reset index to make amino acids a column\n",
        "sns.barplot(\n",
        "    data=aa_df.reset_index(),\n",
        "    x='index',\n",
        "    y='count',\n",
        "    palette='mako'\n",
        ")\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:36.461020Z",
          "iopub.execute_input": "2025-11-11T01:35:36.461442Z",
          "iopub.status.idle": "2025-11-11T01:35:36.752583Z",
          "shell.execute_reply.started": "2025-11-11T01:35:36.461353Z",
          "shell.execute_reply": "2025-11-11T01:35:36.751307Z"
        },
        "id": "4KFmQIdLjJgR"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Number of GO Terms vs Sequence Length"
      ],
      "metadata": {
        "id": "HBwmikgjjJgR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Calculate the number of unique GO terms for each EntryID\n",
        "term_counts = merged_df.groupby('EntryID')['term'].nunique().reset_index(name='num_terms')\n",
        "\n",
        "# Extract unique EntryIDs and their corresponding fasta sequences\n",
        "seq_lengths = merged_df[['EntryID', 'fasta_sequence']].drop_duplicates()\n",
        "\n",
        "seq_lengths['seq_length'] = seq_lengths['fasta_sequence'].str.len()\n",
        "\n",
        "# Merge term counts with sequence lengths based on 'EntryID'\n",
        "entry_stats = pd.merge(term_counts, seq_lengths[['EntryID', 'seq_length']], on='EntryID')\n",
        "\n",
        "# Filter out sequences longer than 5000\n",
        "filtered_stats = entry_stats[entry_stats['seq_length'] <= 5000]\n",
        "\n",
        "plt.figure(figsize=(8,6))\n",
        "\n",
        "# Create a scatter plot of sequence length versus number of terms\n",
        "# 'c' colors the points based on 'num_terms'\n",
        "sc = plt.scatter(\n",
        "    filtered_stats['seq_length'],\n",
        "    filtered_stats['num_terms'],\n",
        "    c=filtered_stats['num_terms'],\n",
        "    cmap='mako',\n",
        "    alpha=0.7\n",
        ")\n",
        "\n",
        "# Add a regression line to show the general trend\n",
        "sns.regplot(\n",
        "    x='seq_length',\n",
        "    y='num_terms',\n",
        "    data=filtered_stats,\n",
        "    scatter=False,\n",
        "    line_kws={'color':'red', 'linewidth':2}\n",
        ")\n",
        "\n",
        "plt.title('Number of GO Terms vs Sequence Length')\n",
        "plt.xlabel('Sequence Length')\n",
        "plt.ylabel('Number of GO Terms')\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-11-11T01:35:36.753731Z",
          "iopub.execute_input": "2025-11-11T01:35:36.754023Z",
          "iopub.status.idle": "2025-11-11T01:35:44.864169Z",
          "shell.execute_reply.started": "2025-11-11T01:35:36.753991Z",
          "shell.execute_reply": "2025-11-11T01:35:44.862789Z"
        },
        "id": "kVvBwf4-jJgR"
      },
      "outputs": [],
      "execution_count": null
    }
  ]
}